Радкевич Марсель

SquirrelLeonid opened the pull request:
@shemloft
Черновой вариант домашки по TagCloud.

SquirrelLeonid committed:
Created projects:

--> TagCloudVisualization
--> TagCloudVisualization_Tests
__________________________________
Implemented some functionality via TDD method
--> CircularCloudLayouter able to put number of  rectangles without intersection. But they will be located diagonally.

SquirrelLeonid committed:
Implemented tag cloud visualizator.

SquirrelLeonid committed:
Added test data generator

SquirrelLeonid committed:
Fixed incorrect positioning of rectangles in the image.

SquirrelLeonid committed:
Minor refactor

Visualizer now also draw boundary and axis.

SquirrelLeonid committed:
Added cloud circle boundary for debugging

SquirrelLeonid committed:
What has benn done:

1) Restructurization. Now, there's 3 modeules:
---->Saving layout to desktop
---->Layouting
---->Visualization

2)Added interfaces for each module

SquirrelLeonid committed:
implemented first steps for circular cloud layout algorithm

-->added class DirectingArrow

SquirrelLeonid committed:
Added length increasing to directing arrow

CircularCloudLayouter.GetCloudRadius now returns radius of an outer circle of cloud.

shemloft commented on cs/TagCloud_TestDataGenerator/Program.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    static class Program
+    {
+        static void Main(string[] args)
```
Кажется это лишнее

shemloft commented on cs/TagCloud_TestDataGenerator/DataGenerator.cs:
```diff
@@ -0,0 +1,29 @@
+    public static class DataGenerator
+    {
+        private const int LowerSize = 15;
+        private const int HigherSize = 150;
```
В данном контексте логичнее звучит `MinSize` и `MaxSize`

shemloft commented on cs/TagCloud_Tests/CircularLayouter_Should.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [Test]
+        public void Have_CenterPoint_AfterCreate()
```
Обычно в подобном именовании тестов есть три части:
Класс(или метод)_ЧтоДолженСделать_Условие. В твоем названии теста здесь, и дальше в паре мест, не нужно первое подчеркивание. Т.е. `HaveCenterPoint_AfterCreate`. Получается законченная фраза с тремя частями `CircularLayouter_ShouldHaveCenterPoint_AfterCreate`.

Вообще, лично мне больше нравится называть класс CircularLayouterTests и в начале названия каждого теста указывать метод, который тестируется, например:
* Ctor_ShouldNotThrow_WhenCreated
* Ctor_ShouldAllowNegativeCoordinatesForCloudCenter
* PutNextRectangle_ShouldReturnSameSize_AfterFirstPut

И т.д.

Оба способа валидны, выбирай какой больше нравится

shemloft commented on cs/TagCloud_Tests/CircularLayouter_Should.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [Test]
+        public void ThrowArgumentException_WhenNotPositiveSize(
```
В этом тесте аж три разных наименования ситуации
* WhenNotPositiveSize
* IncorrectRectangleSizes
* Negative size not allowed

Лучше сойтись на одном и использовать во всех трех местах)

shemloft commented on cs/TagCloud/Layouting/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,126 @@
+            return center;
+        }
+
+        public int GetCloudRadius()
```
Формула здесь неправильная, угол может не попасть (см. картинку)
![image](https://user-images.githubusercontent.com/38860225/140615917-311a990e-676b-443a-b9d2-f0e544e4ad6d.png)

shemloft replied:
Формула не поменялась, ниже дала еще коммент на этот счет

shemloft commented on cs/TagCloud/Layouting/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+
+        private readonly Point center;
```
Можно заменить get-only свойством, которое так же добавить в интерфейс. Не надо будет писать лишние Get-методы. Насчет других Get-методов тоже стоит подумать, не излишни ли они, можно ли заменить свойством в интерфейсе.

shemloft commented on cs/TagCloud/Saving/BitmapToDesktopSaver.cs:
```diff
@@ -0,0 +1,55 @@
+
+        private string GetSaveDirectory()
+        {
+            var desktopPathSuffix = Path.Combine("desktop", "layouts");
```
Падает весь запуск при отсутствии папки layouts, причем с очень неочевидной ошибкой :с
А еще, десктоп не обязательно располагается в папке пользователя.

Да и вообще, лучше не лезть в систему пользователя, а сохранять в папке проекта

shemloft commented on cs/TagCloud/Saving/BitmapToDesktopSaver.cs:
```diff
@@ -0,0 +1,55 @@
+        public BitmapToDesktopSaver()
+        {
+            saveDirectory = GetSaveDirectory();
+            fileNamePrefix = GetFileNamePrefix();
```
Вполне можно сделать просто полем или свойством, зачем делать метод?

shemloft commented on cs/TagCloud/Saving/BitmapToDesktopSaver.cs:
```diff
@@ -0,0 +1,55 @@
+
+        private static void OpenImage(string fullFileName)
+        {
+            Process.Start(fullFileName);
```
`Process` наследуется от `IDisposable`, поэтому на нём должен быть вызван `Dispose`.

shemloft commented on cs/TagCloud/Visualization/CloudDrawer.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class CloudDrawer : ICloudDrawer
+    {
+        public void DrawCloud(Graphics g, Point cloudCenter, List<Rectangle> rectangles)
```
Называется метод DrawCloud, а рисуешь прямоугольники 🤔

shemloft commented on cs/TagCloud/TagCloud.cs:
```diff
@@ -0,0 +1,86 @@
+
+        public void Visualize()
+        {
+            InitCanvasIfNotExist();
```
Использование этого метода нарушает немного повествование кода. Почему вообще ожидается в этих методах (Visualize, VisualizeMarkup), что уже существует canvas. Плюс, в них есть дублирование кода. Кажется, было бы лучше оставить один метод, и ему передавать флагом рисовать или не рисовать разметку

shemloft commented on cs/TagCloud/TagCloud.cs:
```diff
@@ -0,0 +1,86 @@
+            if (canvas != null)
+                return;
+
+            var canvasSize = layouter.GetRectanglesBoundaryBox();
```
TagCloud получается очень непредсказуемым. Например, ты нагенеришь прямоугольников, отрисуешь, а потом решишь нагенерить ещё.. Что будет если ты еще раз отрисуешь?

У тебя получаются два класса, которые хранят состояния (layouter и cloud), причем cloud в какой-то момент сохраняет версию состояния layouter, но никак потом не отслеживает что оно поменялось и т.д.

SquirrelLeonid replied:
@shemloft, я внес небольшие изменения в этом методе, но не уверен, что полностью понял суть комментария.
Теперь данный метод называется `UpdateCanvasSize` заместо `InitCanvasIfNotExist`, и он не проверяет существование поля canvas, а создает новый, вместе с этим обновляя размеры.
При повторной отрисовке изображение заменится новым.

По поводу хранения состояния.
фактически, состоянием раскладки управляет класс `CircularCloudLayouter`, а класс `TagCloud` хранит изображение, но не знает ничего о прямоугольниках ровно до того момента, как потребуется отрисовка.

Действительно ли необходимо отслеживать состояние в `layouter` в реальном времени, если информация о его состоянии требуется только во время отрисовки?

shemloft commented on cs/TagCloud/TagCloud.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            InitCanvasIfNotExist();
+
+            var g = Graphics.FromImage(canvas);
```
Graphics тоже наследуется от `IDisposable`

SquirrelLeonid committed:
1) Made edits to the pull request.

2) Removed IMarkupDrawer and ICloudDrawer. Now there's only IDrawer that combines both of them.

3) Renamed tests

SquirrelLeonid committed:
CircularCloudLayouter was refactored

-Separate ICloudLayouter interface to ICloudLayouter and ILayouter

-DirectingArrow class was taken off
-RectangleExtension class was created

-Class TestDataGenerator renamed to RectangleSizeGenerator

-Added default width and height constants to RectangleSizeGenerator

-Added methods for generating random and fixed sizes to RectangleSizeGenerator

SquirrelLeonid committed:
Refactored TagCloud class

Added coloring for rectangles to Drawer

SquirrelLeonid committed:
Added ReadMe file and layouts examples for 1500 rectangles

SquirrelLeonid committed:
-Added one test in TagCloudTests

-Updated IBitmapSaver interface

SquirrelLeonid committed:
Resharper fixes

SquirrelLeonid committed:
Reduced images size

shemloft commented on cs/tdd.sln:
```diff
@@ -1,11 +1,17 @@
+VisualStudioVersion = 16.0.31729.503
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BowlingGame", "BowlingGame\BowlingGame.csproj", "{AD0F018A-732E-4074-8527-AB2EEC8D0BF3}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BowlingGame", "BowlingGame\BowlingGame.csproj", "{AD0F018A-732E-4074-8527-AB2EEC8D0BF3}"
```
Вообще не очень хорошо, что все проекты TagCloud-а лежат там же, где и проекты с пар. Лучше выделить в отдельный sln

SquirrelLeonid replied:
Я буквально понял первую строку из Readme.md к TagCloud

> Сделай форк этого репозитория. Добавь проект TagsCloudVisualization. Выполняй задания в этом проекте.

shemloft replied:
Ну, ты сделал же в итоге несколько проектов, а не один 🤔

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+        private CircularLayouter layouter;
+
+        [SetUp]
+        public void InitLayouter()
```
Обычно SetUp и TearDown методы так и называют - SetUp и TearDown.

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        [Test]
+        public void Constructor_ShouldAllowNegativeXY_ForCloudCenter()
```
Можно заменить Constructor на Ctor, для краткости, но не обязательно)

SquirrelLeonid replied:
Мне нравится писать полные имена пока они вмещаются в строку и не портят общий вид кода. :)

shemloft commented on cs/TagCloud_TestDataGenerators/DataGenerator.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace TagCloud_TestDataGenerator
+{
+    public static class DataGenerator
```
Проект больше не показывается в sln, но остался в директории, лучше удалить, чтобы не путаться

shemloft commented on cs/TagCloud/Visualization/Drawer.cs:
```diff
@@ -0,0 +1,84 @@
+        public Drawer()
+        {
+            colors = new List<Color>();
+            FillColorList();
```
Кажется, это излишне, можно просто проинициализировать поле сразу:
```
private readonly List<Color> colors = new List<Color>
        {
            Color.Green, 
            Color.Red, 
            Color.CadetBlue // и т.д.
        };
```

shemloft commented on cs/TagCloud/Visualization/Drawer.cs:
```diff
@@ -0,0 +1,84 @@
+
+        public void DrawAxis(Graphics g, Size imgSize, Point cloudCenter)
+        {
+            var cloudCenterOnImg = new Point(cloudCenter.X + imgSize.Width / 2,
```
Считаешь дважды по одной формуле, можно выделить в отдельную функцию

shemloft replied:
Это всё ещё осталось. В каждом использовании этой формулы ты всегда берешь размер и делишь его на два. Можно назвать как-то вроде MovePointToSizeCenter и вот это деление на два оставить в методе

shemloft commented on cs/TagCloud/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,73 @@
+            var boundaryBox = GetRectanglesBoundaryBox();
+            var biggestSide = Math.Max(boundaryBox.Width, boundaryBox.Height);
+
+            return (int)Math.Ceiling(biggestSide * (Math.Sqrt(2) / 2));
```
Этот метод дает не точный радиус, а хотелось бы.. Можно не цепляться за BoundaryBox, а искать самую удаленную от центра точку, и расстояние до нее брать за радиус
Можно сделать красиво через LINQ

shemloft commented on cs/TagCloud/ITagCloud.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        void PutNextTag(Size tagSize);
+
+        string SaveBitmapTo(bool shouldShowLayout, bool shouldShowMarkup, bool openAfterSave);
```
Предлог to в имени метода предполагает, что битмап сохранится в место, указанное в аргументе. Т.к. место не указывается, лучше убрать

shemloft commented on cs/TagCloud/Geometry/RectangleExtension.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public static bool IsIntersectsWithAny(this Rectangle rect, List<Rectangle> rectangles)
+        {
+            foreach (var rectangle in rectangles)
```
Этот метод можно написать одной строкой при помощи LINQ, попробуй)

shemloft commented on cs/TagCloud/Geometry/RectangleExtension.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public static class RectangleExtension
+    {
+        public static bool IsIntersectsWithAny(this Rectangle rect, List<Rectangle> rectangles)
```
Грамматически неверное название, правильно будет или `IntersectsWithAny`, или `IsIntersectingWithAny`

shemloft commented on cs/TagCloud/Visualization/IDrawer.cs:
```diff
@@ -0,0 +1,16 @@
+
+        void DrawCloudBoundary(Graphics g, Size imgSize, Point cloudCenter, int cloudCircleRadius);
+
+        void DrawRectangles(Graphics g, Point cloudCenter, List<Rectangle> rectangles);
```
cloudCenter не используется в методе, решарпер подсказывает

SquirrelLeonid replied:
Странно, мой решарпер ничего не подсказывает. Но действительно не используется, убрал.

shemloft replied:
Мб не настроено.. да и я вообще райдером, а не студией пользуюсь)

shemloft commented on cs/TagCloud/Layouting/ILayouter.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        Size GetRectanglesBoundaryBox();
+
+        List<Rectangle> GetRectanglesCopy();
```
Можно сделать это свойством типа IReadOnlyList или IReadOnlyCollection, не нужно будет писать лишний метод, типа такого:
```
public IReadOnlyList<Rectangle> Rectangles => rectangles.AsReadOnly();
```
Т.к. ты действительно используешь копию, то это не обязательно, но так, полезно знать об этой возможности с:

shemloft commented on cs/TagCloud/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,73 @@
+        {
+            ThrowIfIncorrectSize(rectangleSize);
+
+            var rect = new Rectangle(new Point(Center.X - rectangleSize.Width / 2,
```
Создание этой точки используется здесь, в MoveMiddlePointToCurrentLocation и в Drawer дважды с другим знаком. Можно вынести создание этой точки в метод с говорящим названием куда-нибудь в Extensions, заодно тесты написать.

shemloft commented on cs/TagCloud/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,73 @@
+            {
+                arrow.Rotate();
+                rect.Location = arrow.GetEndPoint();
+                rect.MoveMiddlePointToCurrentLocation();
```
Лучше делать это изменение локации явным, по названию конечно понятно, что метод изменяет сам объект, но лучше отдельно посчитать новую точку, и отдельно в этом методе присвоить новую локацию

shemloft commented on cs/TagCloud/Program.cs:
```diff
@@ -0,0 +1,25 @@
+        private static void Main(string[] args)
+        {
+            var bitmapSaver = new BitmapSaver();
+            var layouter = new CircularLayouter(new Point(0, 0));
```
Можно сделать дефолтный конструктор с центром в нуле

shemloft commented on cs/TagCloud/TagCloud.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            for (var i = 0; i < rectangles.Count; i++)
+            {
+                var newX = rectangles[i].X + imgSize.Width / 2;
```
О, еще использование этой формулы

shemloft commented on cs/TagCloud/TagCloud.cs:
```diff
@@ -0,0 +1,85 @@
+            this.visualizer = visualizer;
+        }
+
+        public string SaveBitmapTo(bool shouldShowLayout, bool shouldShowMarkup, bool openAfterSave)
```
Здесь вполне можно сделать дефолтные значения

shemloft commented on cs/TagCloud_Tests/TagCloudTests.cs:
```diff
@@ -0,0 +1,21 @@
+                null, null);
+
+            Action act = () => tagCloud.SaveBitmapTo(true, true, false);
+            act.Should().Throw<ArgumentException>("image width and height can't be lower than 1");
```
Выглядит так, будто ты хочешь проверить, что именно с этим сообщением выбросилось сообщение, но этот метод так не проверяет

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+
+        [Test]
+        public void Constructor_ShouldSetCenter_WhenCreate()
```
Этот и предыдущий тесты имеют одинаковое тело, лучше собрать в один и точку задавать TestCase-ом

shemloft replied:
А почему ты сделал ValueSource, а не TestCase на каждый вариант?

SquirrelLeonid replied:
TestCase не позволяет передавать структуры в качестве параметра, ругаясь на то, что атрибутом должно быть выражение константного типа.

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+            var rectSize = new Size(10, 10);
+
+            var rect = layouter.PutNextRectangle(rectSize);
+            rect.Location
```
Где-то arrange-act-assert разделены пустыми строчками, где-то (например здесь) нет. Лучше везде разделить

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+            var expectedRatio = 0.8d;
+            var actualRatio = GetDiametersRatio();
+
+            actualRatio.Should().BeGreaterOrEqualTo(expectedRatio);
```
Упал :с 
Еще и картинка не похожа на примеры
![image](https://user-images.githubusercontent.com/38860225/141511033-ce5465e7-0cec-45f1-a67b-ff417ceb0dd2.png)

shemloft replied:
Ну и видно, что радиус все еще неправильно считается

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+                new BitmapSaver(), new Visualizer(new Drawer()));
+
+            var path = tagCloud.SaveBitmapTo(true, true, false);
+            var msg = string.Join("", "Tag cloud visualization saved to file\n", path);
```
Интерполяция строк!!

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+            return smallestSide / biggestSide;
+        }
+
+        private List<Rectangle> PutNextNRectangles(int n)
```
LINQ

shemloft commented on cs/TagCloud_Tests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,181 @@
+        {
+            var context = TestContext.CurrentContext;
+
+            if (!IsPassedLayoutTest(context))
```
Мне не очень нравится, что тут такая проверка. Рисуй для всех тестов (TearDown запускается после каждого теста), кроме случаев, когда нет прямоугольников

shemloft replied:
Проверка все еще осталась

shemloft reviewed (commented):
Хотелось бы тестов и на другие классы, на геометрические точно

shemloft commented:
Добавь еще тест на плотность

SquirrelLeonid committed:
1) RectangleExtension

--> Rename IsIntersectsWithAny method to IntersectsWithAny
--> Rewrote IntersectWithAny to LINQ

2) ITagCloud
--> Renamed method SaveBitmapTo to SaveBitmap

3) IDrawer
--> Removed unused parameter in DrawRectangles method

4) CircularLayouter
--> Added default constructor with Point.Empty

SquirrelLeonid committed:
--> Renamed TestDataGenerator.GenerateNextNFixedSize method

--> updated images for 1500 rectangles

--> fixed formula for cloud radius calculation
----> Now images don't have whitespaces outside the circle

--> Added some methods to RectangleExtension

SquirrelLeonid committed:
Add method to PointExtension that moving Point to specified X and Y

SquirrelLeonid committed:
Separated TagCloud to other solution

SquirrelLeonid committed:
Moved layouts directory to solution root directory

Resized images

SquirrelLeonid committed:
Added tests for classes from Geometry

SquirrelLeonid committed:
Added density test

SquirrelLeonid committed:
Resharper refactor

shemloft commented on cs/TagCloudTask/TagCloudTask/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,81 @@
+        private readonly DirectingArrow _arrow;
+        private readonly List<Rectangle> _rectangles;
+
+        public CircularLayouter()
```
Дублирование конструкторов, придумай как сократить

shemloft commented on cs/TagCloudTask/TagCloudTask/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,81 @@
+                _arrow.Rotate();
+                var arrowEndPoint = _arrow.GetEndPoint();
+                rect.Location = arrowEndPoint;
+                rect = rect.MoveMiddlePointToCurrentLocation();
```
Сделай пересчитывание точки одной функцией, а потом уже складывай ее в Location. Сейчас ты складываешь и внутри пересчитываешь, это не оч

shemloft commented on cs/TagCloudTask/TagCloudTask/Layouting/CircularLayouter.cs:
```diff
@@ -0,0 +1,81 @@
+            return _rectangles.Count == 0
+                ? 0
+                : (int)Math.Ceiling(_rectangles
+                    .Max(rectangle => rectangle.GetLongestDistanceFromPoint(Center)));
```
Попробуй сначала собрать все точки, потом посчитать расстояния, потом выбрать максимальную. Не будет вот этой лишней функции GetLongestDistanceFromPoint

shemloft commented on cs/TagCloudTask/TagCloudTaskTests/CircularLayouterTests.cs:
```diff
@@ -0,0 +1,192 @@
+
+        [Test]
+        public void PutNextRectangle_ShouldThrow_WhenIncorrectSize(
+            [ValueSource(nameof(IncorrectRectangleSizes))]
```
Луше выделить TestCase-ами, каждый со своим названием, так же в CloudCenters

SquirrelLeonid replied:
Здесь аналогичная ситуация к комментарию выше.
TestCase не позволяет передавать структуры в качестве параметра, ругаясь на то, что атрибутом должно быть выражение константного типа.

SquirrelLeonid committed:
Renamed private fields from "_<name>" to "<name>"

Removed GetEndPoint method from Directing Arrow. Now Rotate method returns Point

Added extension method to PointExtension

Removed ExtensionMethod from RectangleExtension

SquirrelLeonid committed:
Removed unused test from RectangleExtensionTestsTests

Rewrite LINQ in GetCloudBoundaryRadius method
